qutip/nonmarkov/bofin.py

```diff
@@ -321,10 +352,24 @@ def __init__(
         self.coup_op, self.ck, self.vk, self.NR, self.NI = (
             _mangle_bath_exponents_bosonic(coup_op, ckAR, ckAI, vkAR, vkAI)
         )
+
         self.progress_bar = BaseProgressBar()
 
+        self.kcut = len(self.coup_op)
+        self.bath = BathStates([self.N_cut + 1] * self.kcut, self.N_cut)
+
         self._configure_solver()
 
+    def _pad_op(self, op, row_he, col_he):
+        """
+        Pad op into its correct position within the larger HEOM liouvillian
+        for the given row and column bath states.
+        """
+        nhe = self.bath.n_states
+        rowidx = self.bath.idx(row_he)
+        colidx = self.bath.idx(col_he)
+        return cy_pad_csr(op, nhe, nhe, rowidx, colidx)
```

Comment:
just a note:  I did a lot of testing on difficult cases, compared to the C++ version in bofin_fast, and this cy_pad_csr stuff is still a big bottleneck.  I saw a significant benefit by going back to the ''insertion'' approach @tehruhn  took, but also making sure that the RHS is in lil format first. In the old notation the insertion was done with 

   # Fill into larger L
    nidx = self.he2idx[he_n]
    block = self.N ** 2
    pos = int(nidx * block)
    self.L_helems[pos : pos + block, pos : pos + block] += L

I guess you could have the grad functions just return the small operator L, and do a similar insertion in the RHS construction function (+starting with RHS in lil format, and then switching back to CSR after everything is constructed).

Reply:
I implement bothed and got the following results on the test suite:

* `pad-csr`: ~11s
* `inplace add, CSR`: ~27s
* `inplace add, LIL`: ~27s
* `inplace add, DOK`: ~22s

The tests use very small systems though, so perhaps we need a bigger system for benchmarking.

The `_rhs` method now has an `_use_pad` option that defaults to `True`. You can set it to `False` if you want to use the `L[..., ...] += op` code though. You can also change `lil_matrix` to `dok_matrix` or `csr_matrix` if you want to experiment with different sparse matrix types.